core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusMetricRequestDecorator.java

```diff
@@ -0,0 +1,370 @@
+     * @return A service decorator function
+     */
+    public static <T extends MetricLabel<T>, I extends Request, O extends Response>
+    PrometheusMetricRequestDecorator<T, I, O> decorateService(
+            CollectorRegistry collectorRegistry,
+            Iterable<T> metricLabels,
+            Function<RequestLog, Map<T, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry,
+                                                      metricLabels,
+                                                      labelingFunction,
+                                                      "armeria_server");
+    }
+
+    /**
+     * Returns a new {@link Client} decorator that tracks request stats using the Prometheus metrics  library.
+     *
+     * @param <I> Request type
+     * @param <O> Response type
+     * @param collectorRegistry Prometheus registry
+     * @param metricLabels Labels names for metrics
+     * @param labelingFunction map of labels to values
+     * @return A client decorator function
+     */
+    public static <T extends MetricLabel<T>, I extends Request, O extends Response>
+    PrometheusMetricRequestDecorator<T, I, O> decorateClient(
+            CollectorRegistry collectorRegistry,
+            T[] metricLabels,
+            Function<RequestLog, Map<T, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry,
+                                                      metricLabels,
+                                                      labelingFunction,
+                                                      "armeria_client");
+    }
+
+    /**
+     * Returns a new {@link Client} decorator that tracks request stats using the Prometheus metrics  library.
+     *
+     * @param <I> Request type
+     * @param <O> Response type
+     * @param collectorRegistry Prometheus registry
+     * @param metricLabels Labels names for metrics
+     * @param labelingFunction map of labels to values
+     * @return A client decorator function
+     */
+    public static <T extends MetricLabel<T>, I extends Request, O extends Response>
+    PrometheusMetricRequestDecorator<T, I, O> decorateClient(
+            CollectorRegistry collectorRegistry,
+            Iterable<T> metricLabels,
+            Function<RequestLog, Map<T, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry,
+                                                      metricLabels,
+                                                      labelingFunction,
+                                                      "armeria_client");
+    }
+
+    private final Function<RequestLog, Map<T, String>> labelingFunction;
+    private final MetricAdapter metricAdapter;
+
+    private PrometheusMetricRequestDecorator(
+            CollectorRegistry collectorRegistry,
+            Object[] metricLabels,
```

Comment:
I see. Thanks! 
yeah Object produces ``Caused by: java.lang.ClassCastException: [Ljava.lang.Object; cannot be cast to [Lcom.linecorp.armeria.common.metric.MetricLabel;``